### DISCUSSION TITLE
Profiling engine events, hydration / casting etc

### DISCUSSION THREAD

**Discussion by Lewiscowles1986** [sqlalchemy/sqlalchemy#11216] [Usage Questions]
Hello 👋 

Back with more awkward questions. Thank you for your patience.

Has anyone got pre-existing work (pypi package, or tips) on hooking into the event system, to gain visibility over the cost of some events, tooling and things like constructing `Row` and `BaseModel` derrived objects from the underlying engine?

I'm getting some PyMySQL and AIO MySQL parts instrumented for free by AWS Lambda, but sometimes there are multi millisecond gaps (even 10's of ms) between `with sessionmaker(engine):` or `with Session(engine):` connecting I'd assume, and then running a query, and marshalling results.

This isn't urgent, but I was wondering how much instrumentation is there to be used, and if this is a solved problem for the community?

This is even true of simple queries, and to a lesser extent is true of lower-level primitives such as Connection and cursor.

**Comment by CaselIT** [comment]
Hi,

It's not very clear what you are looking for.
Are you looking for a way of profiling sqlalchemy code? Of so there is this paga on the docs that should be a good starting point https://docs.sqlalchemy.org/en/20/faq/performance.html#how-can-i-profile-a-sqlalchemy-powered-application

**Reply by Lewiscowles1986** [reply]
Thanks @CaselIT I am indeed currently using AWS X-Ray to instrument this level of profiling when deployed on AWS; Maybe there is no in-between, but I am using Profiler, as well as manual instrumentation.

What I see is that the connection aquiring and closing takes time, but even with that spied upon, there is still time going missing; which seems to be missing locally; so I was looking for ideally a Pypi package, or perhaps links to blog posts of what others might be doing.

The docs are good, but they are a little... Basic.

**Reply by CaselIT** [reply]
that's what there is on the sqlalchemy side...
